state/state_test.go

```diff
@@ -5225,3 +5225,29 @@ func (s *StateSuite) TestPeerRelationCreatesApplicationSettings(c *gc.C) {
 	_, err = settings.ReadSettings(key)
 	c.Assert(err, jc.ErrorIsNil)
 }
+
+func (s *StateSuite) TestAddRelationAssertsRelationCount(c *gc.C) {
+	f := factory.NewFactory(s.State, s.StatePool)
+	wordpressCharm := f.MakeCharm(c, &factory.CharmParams{Name: "wordpress"})
+	f.MakeApplication(c, &factory.ApplicationParams{Name: "wordpress", Charm: wordpressCharm})
+
+	mysqlCharm := f.MakeCharm(c, &factory.CharmParams{Name: "mysql"})
+	f.MakeApplication(c, &factory.ApplicationParams{Name: "mysql", Charm: mysqlCharm})
+
+	eps, err := s.State.InferEndpoints("wordpress", "mysql")
+	c.Assert(err, gc.Equals, nil)
+
+	defer state.SetBeforeHooks(c, s.State, func() {
+		// we modify the relationcount of the wordpress application as if
+		// another relation was added in the meantime and we expect the
+		// AddRelation call to fail because an assertion for relationcount
+		// was added.
+		s.applications.Update(
+			bson.D{{"_id", state.DocID(s.State, "wordpress")}},
+			bson.D{{"$inc", bson.D{{"relationcount", 1}}}},
+		)
+	}).Check()
+
+	_, err = s.State.AddRelation(eps...)
+	c.Assert(err, gc.ErrorMatches, `cannot add relation \"wordpress:db mysql:server\": state changing too quickly; try again soon`)
```

Comment:
`state changing too quickly` means we couldn't make the asserts work for success within 3 attempts and it's a coding failure. The first attempt would fail, but we should be able to revaluate and succeed on a 2nd attempt. 

If AddRelation should fail because the application/endpoint combinations cannot be made a 2nd time, then the error should be explicit, not `state changing too quickly`.

Reply:
This is explicitly testing the assertions always fail because of the relation count. See the `SetBeforeHooks` call above.